### PR TITLE
Optional column metadata field to store bounding box information

### LINKED ISSUE #8: Store optional bounding box information in the column metadata
There was a bit of discussion around this in https://github.com/opengeospatial/cdw-geo/pull/4. 

The proposal is to add an _optional_ column metadata field (alongside the currently required "crs" and "encoding" fields) that describes the bounding box of the full file (so the overall bounding box or envelope of all geometries in the file). 

In the geo-arrow-spec version of this metadata specification, we are already using it (https://github.com/geopandas/geo-arrow-spec/blob/main/metadata.md#bounding-boxes), and there it takes the form of a a list that specifies the minimum and maximum values of each dimension. So for 2D data it would look like ``"bbox" : [<xmin>, <ymin>, <xmax>, <ymax>]``.

This formatting aligns with for example the GeoJSON spec (https://datatracker.ietf.org/doc/html/rfc7946#section-5).

---

This optional information can be useful when processing this data. For example, in [dask-geopandas](https://github.com/geopandas/dask-geopandas) we already make use of this feature to filter partitions (sub-datasets) of a dataset. When using Parquet, people often make use of "partitioned datasets", where the dataset consists of (potentially nested directories of) many smaller Parquet files. In such a situation, you could spatially sort the data when dividing into partitions and each individual file could contain the data of a certain region. If each individual Parquet file would then store information about the bounding box of their geometries, this allows to only read those files needed when doing a spatial query while reading the dataset (a kind of "predicate pushdown", as can be done for Parquet based on column statistics).

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #8

### CODE REVIEW

**Comment by cholmes** [comment]
> Do we want to relate this to https://github.com/opengeospatial/geoparquet/issues/9?

I'm not sure exactly how we relate it, but I do think we should explicitly mention the interaction between that geodesic=true/false and the bounding box - if that field changes the interpretation of anything here.

> It mentions how to deal with the poles, by using some placeholder values like minlat. 

Wow, I've read that spec a number of times and never even noticed this. I feel ok with just referencing the spec and not trying to clarify this particular issue. If someone actually has a problem with it then we could redo it. But I think it's nice to be able to point at something else that has lots of parsers written for it, instead of making people carefully read ours and try to do something slightly different.

**Comment by kylebarron** [comment]
> Wow, I've read that spec a number of times and never even noticed this

👍 , just saw this for the first time!

> * It mentions how to deal with the poles, by using some placeholder values like `minlat`. However, it's not clear to me what this is (is this a string in JSON? But it's not formatted as a string). I am also not aware of an implementation actually doing this (for example, GDAL does not support this AFAIK)

My read of this is: in each of those examples, `minlat`, `maxlat`, `eastlon`, `westlon` aren't string literals intending to be saved to JSON but rather variables that you would replace with numbers for your own bbox. For example

> A bounding box that contains the North Pole extends from a southwest
>    corner of "minlat" degrees N, 180 degrees W to a northeast corner of
>    90 degrees N, 180 degrees E.  Viewed on a globe, this bounding box
>    approximates a spherical cap bounded by the "minlat" circle of
>    latitude.
> 
>    "bbox": [-180.0, minlat, 180.0, 90.0]

My read of that is that if you had a bounding box to cover the globe above `70` degress latitude, your bounding box would be

```json
[-180.0, 70, 180.0, 90.0]
```

**Comment by jorisvandenbossche** [comment]
> My read of this is: in each of those examples, `minlat`, `maxlat`, `eastlon`, `westlon` aren't string literals intending to be saved to JSON but rather variables that you would replace with numbers for your own bbox.

Indeed! By chatting with @paleolimbot about this, he mentioned the same. I completely missed this from an initial reading, but that makes sense now.

So to follow up on my own questions:

> It mentions how to interpret it in case of antimeridian crossing -> this seems good to follow.

That is consistent with how PROJ or S2 specify bounding boxes / areas (see also the comment at https://github.com/opengeospatial/geoparquet/issues/8#issuecomment-1055557211). 
I think it is fine to follow this, but implementors need to be aware that this makes interpreting the bbox more complex (you need to check for the east (ymax ) value being smaller than the west (ymin) value, and handle this case properly for crossing the antimeridian). For example, in the current GeoPandas implementation of this, we just make a rectangular polygon of those 4 values, and this gives the wrong polygon (not crossing the antimeridian, but rather the complementary bbox).

> It mentions how to interpret the connecting lines (the edges of the box). Do we want to relate this to https://github.com/opengeospatial/geoparquet/issues/9?

This is actually fine, and isn't influenced by the `edges="planar"|"spherical"` flag. Because of the way how a bbox is defined (take the min/max range of coordinates in each dimension), it can in practice always be considered as planar.

> It mentions how to deal with the poles, by using some placeholder values like `minlat`.

This is clarified above. 

---

In summary, I am now convinced it is good to specify this field follows the GeoJSON specification.

File: format-specs/geoparquet.md [hunk 1]
```diff
@@ -78,6 +79,16 @@ This is the binary format that the geometry is encoded in. The string 'WKB' to r
 [Well Known Binary](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary) is the only current option, but future versions
 of the spec may support alternative encodings. This should be the ["standard"](https://libgeos.org/specifications/wkb/#standard-wkb) WKB representation.
 
+#### bbox
+
+Bounding boxes are used to help define the spatial extent of each geometry column.
+Implementations of this schema may choose to use those bounding boxes to filter
+partitions (files) of a partitioned dataset.
```
[the remark this answers is not in the document]
Comment (cholmes) [reply format-specs/geoparquet.md:86]: I think that could be nice to have, but I think it's also ok to ship 0.1 without fully describing how to do the spatial partitioning. I feel like partitioning is not the only reason to have a bbox - it's nice to be able to get the bounds without iterating through every feature. I see this as just establishing the field, and in 0.2 we'll tackle spatial indexing as a whole, with #13.

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply format-specs/geoparquet.md:86]: A description like that would be useful, but doesn't necessarily belong here in the `bbox` specification, I think (especially if we want to go a bit more in depth). 
The main reason I mentioned it here was give _some_ context of why it could be useful. But I could also leave it out alltogether?

> 2\. and the bounding box of the root / common_metadata file would be the unary union of the bounding boxes (right?).

That's not something that is currently guaranteed, at least not in the spec (this is just about the bbox field for a single parquet file), and we also don't do that in the geopandas implementation actually. But so that's a good question ..

Given that a proper spatial partitioning is quite complex (eg what to do with the `_metadata` file if present), and we might want to promote other ways than a bbox in each individual file (that is not very efficient as you need to read the footer of each file to build up your spatial partitioning structure), I would indeed leave that for a separate discussion (and for 0.2), and focus here just on adding a bbox for individual files (and as @cholmes says, that can indeed be useful in general, not only for spatial partitioning.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply format-specs/geoparquet.md:86]: Indeed. Let's keep it simple for now. A hint as to why it might be useful, like you have now, is better than nothing IMO.
